src/Nethermind/Nethermind.Serialization.Rlp/Rlp.cs

```diff
@@ -1608,74 +1590,28 @@ public static int LengthOfNonce(ulong _)
 
         public static int LengthOf(long value)
         {
-            // everything has a length prefix
-            if (value < 0)
+            if ((ulong)value < 128)
             {
-                return 9;
+                return 1;
             }
-
-            if (value < 256L * 256L * 256L * 256L * 256L * 256L * 256L)
+            else
             {
-                if (value < 256L * 256L * 256L * 256L * 256L * 256L)
-                {
-                    if (value < 256L * 256L * 256L * 256L * 256L)
-                    {
-                        if (value < 256L * 256L * 256L * 256L)
-                        {
-                            if (value < 256 * 256 * 256)
-                            {
-                                if (value < 256 * 256)
-                                {
-                                    if (value < 128)
-                                    {
-                                        return 1;
-                                    }
-
-                                    return value < 256 ? 2 : 3;
-                                }
-
-                                return 4;
-                            }
-
-                            return 5;
-                        }
-
-                        return 6;
-                    }
-
-                    return 7;
-                }
-
-                return 8;
+                // everything has a length prefix
+                return 1 + sizeof(ulong) - (BitOperations.LeadingZeroCount((ulong)value) / 8);
             }
-
-            return 9;
         }
 
         public static int LengthOf(int value)
         {
-            // everything has a length prefix
-            if (value < 0)
+            if ((uint)value < 128)
             {
-                return 9; // we cast it to long now
+                return 1;
```

Comment:
I just realized there's behavioral difference here for negative 32-bit values. Why is this being casted to long and is there anything that depends on this? 🤔

Reply:
Seems like it really matters, after all. `RlpStream.Encode(int)` delegates to `RlpStream.Encode(long)` which to me is a bit bizarre. 

Should restore old length behaviour for the negative 32-bit integers or fix `RlpStream`?

Reply:
Rlp stores positive integers up to 2^64; so `ulong` size, guessing negative `int` is sign extension to `long` then flip to `ulong`ish. First bit on a negative number will be 1 and first byte will have content so it can only be max size, no leading zeros to skip?

Reply:
Yeah not sure there is such a thing so it would be encoded as `(ulong)(long)x`

Reply:
Hmm, could we more generally delegate all `int` encoding methods to the `long` ones then?

Reply:
Yes that would likely work, if `int` was passed in to `long` (rather than pre-sign conversion to `uint`)